src/opm/parser/eclipse/EclipseState/Grid/EclipseGrid.cpp

```diff
@@ -331,18 +521,436 @@ namespace Opm {
+                    zcorn[zind + 1] = DEPTHZ[i+j*(dims[0]+1) +1 ] + z + DZV[k];
+
+                    zind = zind + dims[0]*2;
+
+                    zcorn[zind] = DEPTHZ[i+(j+1)*(dims[0]+1)] + z + DZV[k];
+                    zcorn[zind + 1] = DEPTHZ[i+(j+1)*(dims[0]+1) + 1] + z + DZV[k];
+                }
+            }
+        }
+
+        return zcorn;
+    }
+
+    std::vector<double> EclipseGrid::makeCoordDxDyDzTops(const std::array<int, 3>& dims , const std::vector<double>& DX, const std::vector<double>& DY, const std::vector<double>& DZ, const std::vector<double>& TOPS) {
+
+        std::vector<double> coord;
+        coord.reserve((dims[0]+1)*(dims[1]+1)*6);
+
+        for (int j = 0; j < dims[1]; j++) {
+
+            double y0 = 0;
+            double zt = TOPS[0];
+            double zb = zt + sumDZ(0, 0, dims, DZ);
+
+            if (j == 0) {
+                double x0 = 0.0;
+
+                coord.push_back(x0);
+                coord.push_back(y0);
+                coord.push_back(zt);
+                coord.push_back(x0);
+                coord.push_back(y0);
+                coord.push_back(zb);
+
+                for (int i = 0; i < dims[0]; i++) {
+
+                    size_t ind = i+j*dims[0]+1;
+
+                    if (i == (dims[0]-1)) {
+                        ind=ind-1;
+                    }
+
+                    zt = TOPS[ind];
+                    zb = zt + sumDZ(i, j, dims, DZ);
+
+                    double xt = x0 + DX[i+j*dims[0]];
+                    double xb = sumDX(j, dims[2]-1, i, dims, DX);
+
+                    coord.push_back(xt);
+                    coord.push_back(y0);
+                    coord.push_back(zt);
+                    coord.push_back(xb);
+                    coord.push_back(y0);
+                    coord.push_back(zb);
+
+                    x0=xt;
+                }
+            }
+
+            int ind = (j+1)*dims[0];
+
+            if (j == (dims[1]-1) ) {
+                ind = j*dims[0];
+            }
+
+            double x0 = 0.0;
+
+            double yt = sumDY(0, 0, j, dims, DY);
+            double yb = sumDY(0, dims[2]-1, j, dims, DY);
+
+            zt = TOPS[ind];
+            zb = zt + sumDZ(0, j, dims, DZ);
+
+            coord.push_back(x0);
+            coord.push_back(yt);
+            coord.push_back(zt);
+            coord.push_back(x0);
+            coord.push_back(yb);
+            coord.push_back(zb);
+
+            for (int i=0; i < dims[0]; i++) {
+
+                ind = i+(j+1)*dims[0]+1;
+
+                if (j == (dims[1]-1) ) {
+                    ind = i+j*dims[0]+1;
+                }
+
+                if (i == (dims[0] - 1) ) {
+                    ind=ind-1;
+                }
+
+                zt = TOPS[ind];
+                zb = zt + sumDZ(i, j, dims, DZ);
+
+                double xt=-999;
+                double xb=-999;
+
+                if (j == (dims[1]-1) ) {
```

Comment:
Consider if the sumDX() and sumDY() functions here should be refactored similar to the use of `partial_sum()` suggested previously. Could they compute and store the results for a whole row/column in a reasonable way? I have not done any benchmarking, but I suspect that for really large grids, this could become a performance bottleneck.

Reply:
These functions are only used with cartesian grids, and these grid type is not commonly used. At least for field models. I'm not aware of any models which uses this type of grids except for the SPE models. I recommend that we look into this when we have a use case for this.

Reply:
Ok, I agree it is probably not used very much.